### DISCUSSION TITLE
Make traits composable when they are used

### DISCUSSION THREAD

**Discussion by duckki** [modular/modular#1551] [Ideas]
The current (Mojo 0.6.1) trait system forces to pick just one trait for any given type parameter, not allowing to specify multiple traits.

Example:
```mojo
# OK
fn useStringable[T: Stringable](x: T): ...

# Not easy.
fn useStringableCollectionElement[T: ????](x: T): ...
# We need `StringableCollectionElement` trait!
```

`CollectionElement` inherits multiple traits and still there is feature request for `ComparableCollectionElement` (https://github.com/modularml/mojo/issues/1526). Similarly, someone will want `HashableCollectionElement` and someone will want `StringableCollectionElement` and `StringableComparableCollectionElement` and so forth. Each collection implementation will have to choose one of ever increasing set of traits. The trait system will face the same consequence of C++/Java/C# standard library having to define a huge hierarchy of trait inheritance where any possible combinations of traits must be defined ahead of time.

On the other hand, Lean allows multiple traits (typeclass instances) for a given type parameter (https://lean-lang.org/lean4/doc/typeclass.html).

In Lean syntax:
```lean
def useStringable {T: Type} [Stringable T] (x: T) (y: T) :=
    ...
def useStringableCollectionElement {T: Type} [Stringable T] [CollectionElement T] (x: T) (y: T) :=
    ...
```

Translated to imaginary Mojo syntax:
```mojo
fn useStringable[T: AnyType, Stringable(T)](x: T, y: T):
    ...

fn useStringableCollectionElement[T: AnyType, Stringable(T), CollectionElement(T)](x: T, y: T):
    ...
```

#### Extensions
This idea is orthogonal to the extension mechanism. However, extension will be the most important use case and I want to show the idea applies to it as well.

Example: We'd like make `DynamicVector[T]` to be stringable, whenever `T` is Stringable & CollectionElement. And we don't want to create StringableDynamicVector for that or make all DynamicVectors to be Stringable.

In Lean syntax,
```lean
instance {T: Type} [Stringable T] [CollectionElement T] : Stringable (DynamicVector T) :=
    ...
```

Translated to imaginary Mojo syntax:
```
extension [T: AnyType, Stringable(T), CollectionElement(T)] Stringable(DynamicVector[T]):
   ...
```

#### Other languages
Rust and Haskell allow to compose traits as well with different syntax.

In Rust:
```rust
impl <A: TraitB + TraitC, D: TraitE + TraitF> MyTrait<A, D> for YourType {}

// Expressing bounds with a `where` clause
impl <A, D> MyTrait<A, D> for YourType where
    A: TraitB + TraitC,
    D: TraitE + TraitF {}
```
(from https://doc.rust-lang.org/rust-by-example/generics/where.html)

In Haskell:
```haskell
compareWithHundred :: (Num a, Ord a) => a -> Ordering  
compareWithHundred x = compare 100 x  

instance VarMonad (ST s) (STRef s) where ...
```
(from https://wiki.haskell.org/Multi-parameter_type_class)

Rust-style composition syntax would work with Mojo.

**Comment by YichengDWu** [comment]
Related to #1431.
